blockchain/interceptors.go

```diff
@@ -20,79 +22,183 @@ const (
-	if !ok {
-		return status.Errorf(codes.InvalidArgument, "missing metadata")
+// Payment represents payment handler specific data which is validated
+// and used to complete payment.
+type Payment interface{}
+
+// PaymentHandler interface which is used by gRPC interceptor to get, validate
+// and complete payment. There are two payment handler implementations so far:
+// jobPaymentHandler and escrowPaymentHandler.
+type PaymentHandler interface {
+	// Payment extracts payment data from gRPC request context.
+	Payment(context *GrpcStreamContext) (payment Payment, err *status.Status)
+	// Validate checks validity of payment data, it returns nil if data is
+	// valid or appropriate gRPC error otherwise.
+	Validate(payment Payment) (err *status.Status)
+	// Complete completes payment if gRPC call was successfully proceeded by
+	// service.
+	Complete(payment Payment) (err *status.Status)
+	// CompleteAfterError completes payment if service returns error.
+	CompleteAfterError(payment Payment, result error) (err *status.Status)
+}
+
+// GrpcStreamInterceptor returns gRPC interceptor to validate payment. If
+// blockchain is disabled then noOpInterceptor is returned.
+func GrpcStreamInterceptor(processor *Processor) grpc.StreamServerInterceptor {
+	if !processor.enabled {
+		log.Info("Blockchain is disabled: no payment validation")
+		return noOpInterceptor
 	}
 
-	paymentHandler, err := p.newPaymentHandlerByType(md)
+	log.Info("Blockchain is enabled: instantiate payment validation interceptor")
+	interceptor := &paymentValidationInterceptor{
+		processor:            processor,
+		jobPaymentHandler:    newJobPaymentHandler(processor),
+		escrowPaymentHandler: newEscrowPaymentHandler(processor, nil, nil),
+	}
+	return interceptor.intercept
+
+}
+
+type paymentValidationInterceptor struct {
+	processor            *Processor
+	jobPaymentHandler    *jobPaymentHandler
+	escrowPaymentHandler *escrowPaymentHandler
+}
+
+func (interceptor *paymentValidationInterceptor) intercept(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	var err *status.Status
+
+	context, err := getGrpcContext(ss, info)
+	if err != nil {
+		return err.Err()
+	}
+	log.WithField("context", context).Debug("New gRPC call received")
+
+	paymentHandler, err := interceptor.getPaymentHandler(context)
+	if err != nil {
+		return err.Err()
+	}
+
+	payment, err := paymentHandler.Payment(context)
 	if err != nil {
-		return err
+		return err.Err()
 	}
+	log.WithField("payment", payment).Debug("New payment received")
 
-	err = paymentHandler.validatePayment()
+	err = paymentHandler.Validate(payment)
 	if err != nil {
-		return err
+		return err.Err()
+	}
+	log.Debug("Payment validated")
+
+	e := handler(srv, ss)
+	if e != nil {
+		log.WithError(e).Warn("gRPC handler returned error")
+		err = paymentHandler.CompleteAfterError(payment, e)
+		if err != nil {
+			return err.Err()
+		}
+		return e
```

Comment:
If CompleteAfterError() succeeded does it necessary  to throw error after it?
If yes, could CompleteAfterError() always returns an error (new one or original) so we always returns it from the  CompleteAfterError() method call?

Reply:
Yes it is necessary, because ```e``` and ```err``` are two different errors here. ```CompleteAfterError``` is method which should cleanup payment state after call handler returns. If it succeeds we still should notify service caller that his request ended with error. If it ```CompleteAfterError``` returns error in turn we have a choice to return ```e``` or ```err```. Actually after thinking about it twice I think that it is better still return ```e```, but log ```err``` instead.